File: packages/dotnet/src/AxeneMailerClient.cs
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Axene.Mailer.Internal;

namespace Axene.Mailer
{
    /// <summary>
    /// Official .NET client for Axene Mailer: send receipts, confirmations, and
    /// campaigns from your own domain. Priced in KES, billed via M-Pesa.
    /// </summary>
    /// <example>
    /// <code>
    /// var axene = new AxeneMailerClient("axm_k_your_api_key");
    /// var res = await axene.SendAsync(new SendEmail {
    ///     From = "hello@yourdomain.com",
    ///     To = { "customer@example.com" },
    ///     Subject = "Your receipt",
    ///     Html = "&lt;p&gt;Thanks for your order.&lt;/p&gt;"
    /// });
    /// </code>
    /// </example>
    public sealed class AxeneMailerClient : System.IDisposable
    {
        private readonly ApiTransport _transport;

        /// <summary>Create a client.</summary>
        /// <param name="apiKey">API key from your dashboard (starts with <c>axm_k_</c>).</param>
        /// <param name="baseUrl">Override the API base URL. Defaults to https://mail.axene.io.</param>
        /// <param name="httpClient">Supply your own <see cref="HttpClient"/> (recommended under DI). One is created and owned otherwise.</param>
        /// <param name="maxRetries">Total attempts on 429 / 5xx. Defaults to 3.</param>
        public AxeneMailerClient(string apiKey, string? baseUrl = null, HttpClient? httpClient = null, int maxRetries = 3)
        {
            _transport = new ApiTransport(apiKey, baseUrl, httpClient, maxRetries);
        }

        /// <summary>Send a single email.</summary>
        public Task<SendEmailResult> SendAsync(SendEmail email, CancellationToken ct = default)
            => _transport.RequestAsync<SendEmailResult>(HttpMethod.Post, "v1/emails/", email.ToWire(), ct);

        /// <summary>Send up to your plan's batch limit in one call. The API accepts a bare array of messages.</summary>
        public Task<BatchResult> SendBatchAsync(IEnumerable<SendEmail> emails, CancellationToken ct = default)
            => _transport.RequestAsync<BatchResult>(HttpMethod.Post, "v1/emails/batch",
                emails.Select(e => e.ToWire()).ToList(), ct);

        /// <summary>Fetch a single email and its current status.</summary>
        public Task<EmailRecord> GetAsync(string id, CancellationToken ct = default)
            => _transport.RequestAsync<EmailRecord>(HttpMethod.Get, $"v1/emails/{System.Uri.EscapeDataString(id)}", null, ct);

        /// <summary>
        /// Dry-run a send: check whether <paramref name="message"/> would be accepted
        /// (sender registered, domain verified, plan limits, account not restricted)
        /// without actually sending it.
        /// </summary>
        public Task<ValidationResult> ValidateAsync(SendEmail message, CancellationToken ct = default)
            => _transport.RequestAsync<ValidationResult>(HttpMethod.Post, "v1/emails/validate", message.ToWire(), ct);

        /// <summary>List your sending domains and their verification status.</summary>
        public Task<List<DomainRecord>> ListDomainsAsync(CancellationToken ct = default)
            => _transport.RequestAsync<List<DomainRecord>>(HttpMethod.Get, "v1/domains/", null, ct);

        /// <summary>Dispose the underlying <see cref="HttpClient"/> if this client created it.</summary>
        public void Dispose() => _transport.Dispose();
    }
}

File: packages/java/src/main/java/io/axene/mailer/AxeneMailerClient.java
package io.axene.mailer;

import java.time.Duration;
import java.util.List;

/**
 * Official Java client for Axene Mailer.
 *
 * <p>Resources are exposed as accessor methods that share a single transport:
 * {@code client.emails()}, {@code client.domains()}, {@code client.contacts()},
 * {@code client.suppressions()}, {@code client.templates()},
 * {@code client.webhooks()}.
 *
 * <pre>{@code
 * AxeneMailerClient axene = new AxeneMailerClient("axm_k_your_api_key");
 * SendEmailResult res = axene.emails().send(SendEmail.builder()
 *     .from("hello@yourdomain.com", "Your Shop")
 *     .to("customer@example.com")
 *     .subject("Your receipt")
 *     .html("<p>Thanks for your order.</p>")
 *     .build());
 * }</pre>
 */
public final class AxeneMailerClient {

    private static final String DEFAULT_BASE = "https://mail.axene.io";

    private final Emails emails;
    private final Domains domains;
    private final Contacts contacts;
    private final Suppressions suppressions;
    private final Templates templates;
    private final Webhooks webhooks;

    /** @param apiKey API key from your dashboard (starts with {@code axm_k_}). */
    public AxeneMailerClient(String apiKey) {
        this(apiKey, DEFAULT_BASE, 3);
    }

    /**
     * @param apiKey     API key from your dashboard.
     * @param baseUrl    override the API base URL (defaults to https://mail.axene.io).
     * @param maxRetries total attempts on 429 / 5xx (defaults to 3).
     */
    public AxeneMailerClient(String apiKey, String baseUrl, int maxRetries) {
        if (apiKey == null || apiKey.trim().isEmpty()) {
            throw new IllegalArgumentException("apiKey is required");
        }
        String base = (baseUrl == null ? DEFAULT_BASE : baseUrl).replaceAll("/+$", "");
        ApiTransport transport = new ApiTransport(apiKey, base, Math.max(1, maxRetries), Duration.ofSeconds(30));
        this.emails = new Emails(transport);
        this.domains = new Domains(transport);
        this.contacts = new Contacts(transport);
        this.suppressions = new Suppressions(transport);
        this.templates = new Templates(transport);
        this.webhooks = new Webhooks(transport);
    }

    /** Send, search, schedule, and inspect emails. */
    public Emails emails() {
        return emails;
    }

    /** Register, verify, and transfer sending domains. */
    public Domains domains() {
        return domains;
    }

    /** Manage subscriber lists, contacts, and bulk sends. */
    public Contacts contacts() {
        return contacts;
    }

    /** Manage the do-not-send suppression list. */
    public Suppressions suppressions() {
        return suppressions;
    }

    /** Manage reusable email templates. */
    public Templates templates() {
        return templates;
    }

    /** Manage event webhooks and inspect deliveries. */
    public Webhooks webhooks() {
        return webhooks;
    }

    // -- convenience shortcuts (delegate to resources) ---------------------

    /** Shortcut for {@code emails().send(...)}. */
    public SendEmailResult send(SendEmail email) {
        return emails.send(email);
    }

    /** Shortcut for {@code emails().sendBatch(...)}. */
    public BatchResult sendBatch(List<SendEmail> emails) {
        return this.emails.sendBatch(emails);
    }

    /** Shortcut for {@code emails().validate(...)}. */
    public ValidationResult validate(SendEmail message) {
        return emails.validate(message);
    }

    /** Shortcut for {@code domains().list()}. */
    public List<DomainRecord> listDomains() {
        return domains.list();
    }
}
